pkg/dnscache/dnscache.go

```diff
@@ -0,0 +1,248 @@
+	ipToDNS      *expirable.LRU[string, *set.Set[string]]
+	dnsToRecords *expirable.LRU[string, dnsRecords]
+
+	rwLock *sync.RWMutex
+}
+
+type Config struct {
+	Enable    bool
+	CacheSize int
+	TTL       int
+}
+
+func New(cfg Config) *DNSCache {
+	ttl := time.Second * time.Duration(cfg.TTL)
+	return &DNSCache{
+		ipToDNS:      expirable.NewLRU[string, *set.Set[string]](cfg.CacheSize, nil, ttl),
+		dnsToRecords: expirable.NewLRU[string, dnsRecords](cfg.CacheSize, nil, ttl),
+		rwLock:       &sync.RWMutex{},
+	}
+}
+
+// Add adds the DNS response into the DNS' LRU cache.
+func (dc *DNSCache) Add(dnsResponse trace.DnsResponseData) error {
+	switch dnsResponse.QueryData.QueryType {
+	case "A":
+	case "AAAA":
+	case "PTR":
+	default:
+		return nil // Do not handle other query types
+	}
+
+	dc.rwLock.Lock()
+	defer dc.rwLock.Unlock()
+
+	query := dnsResponse.QueryData.Query
+	answers := extractAnswers(dnsResponse.DnsAnswer)
+	if isReverseQuery(dnsResponse.QueryData.QueryType) {
+		return dc.addReverseQuery(query, answers)
+	}
+
+	// Check if CNAME entry exists.
+	for i, answer := range answers {
+		switch strings.ToUpper(dnsResponse.DnsAnswer[i].Type) {
+		case "CNAME":
+			if err := dc.addDnsCname(query, answer); err != nil {
+				logger.Errorw("Error adding DNS CName record to DNS Cache", "error", err)
+			}
+			goto end
+		case "ALIAS":
+			if err := dc.addDnsAlias(query, answer); err != nil {
+				logger.Errorw("Error adding DNS CName record to DNS Cache", "error", err)
+			}
+			goto end
+		}
+	}
+
+end:
+	return dc.addDnsQuery(query, answers)
+}
+
+// GetIPsForDNS returns IPs by passed DNS from the LRU cache.
+func (dc *DNSCache) GetIPsForDNS(dns string) ([]string, error) {
+	dc.rwLock.RLock()
+	defer dc.rwLock.RUnlock()
+
+	ipList, ok := dc.dnsToRecords.Get(dns)
+	if !ok {
+		return nil, detect.ErrDataNotFound
+	}
+
+	return ipList.addresses.Items(), nil
+}
+
+// GetIPsForDNS returns IPs by passed DNS from the LRU cache.
+func (dc *DNSCache) GetDNSRecords(dns string) (dnsRecords, error) {
+	dc.rwLock.RLock()
+	defer dc.rwLock.RUnlock()
+
+	records, ok := dc.dnsToRecords.Get(dns)
+	if !ok {
+		return dnsRecords{}, detect.ErrDataNotFound
+	}
+
+	return records, nil
+}
+
+// GetDNSForIP returns DNSs by passed IP from the LRU cache.
+func (dc *DNSCache) GetDNSForIP(ip string) ([]string, error) {
+	dc.rwLock.RLock()
+	defer dc.rwLock.RUnlock()
+
+	dnsList, ok := dc.ipToDNS.Get(ip)
+	if !ok {
+		return nil, detect.ErrDataNotFound
+	}
+
+	return dnsList.Items(), nil
+}
+
+func (dc *DNSCache) addDnsCname(query string, answer string) error {
+	//
+	// Note: Do not use any locks here.
+	//       It is internal function.
+	//       Only the caller (from exposed function must use locking).
+
+	if records, ok := dc.dnsToRecords.Get(query); !ok {
+		dc.dnsToRecords.Add(query, dnsRecords{
+			addresses: set.New[string](),
+			cname:     answer,
+		})
+	} else {
+		// DNS CNAME should always come first in the set
+		records.cname = answer
```

Comment:
I believe this is ok as CNAMEs need to be unique (can't have one domain name pointing to two CNAMEs). The CNAME change would mean the domain was reconfigured (And its okay for us to overwrite the value).